### ISSUE TITLE
Wrap the propTypes with NODE_ENV !== "production"?

### ISSUE THREAD

**Issue by oliviertassinari** [react-bootstrap/react-bootstrap#2101]
### The issue

When consuming `react-boostrap` from the npm registry we are most likely shipping a bundle full of `propTypes` definitions to our end users.
The new [warning](https://github.com/facebook/react/issues/7131) introduced in React v15.3.0 make it clear. Those `propTypes` are only here for the development mode. Sending them down the wire is wasteful.
Empirically, I think that this is responsible for 20% of the `react-boostrap` size.
### The proposed solution

For the npm released version, we could be wrapping all the `propTypes` definitions around the following block:

``` js
process.env.NODE_ENV !== "production"
```

However, that should probably be considered as a breaking change. Users might use them internally even though, I don't think that it's not supposed to be a public API.
### In practice

This is what https://github.com/callemall/material-ui/pull/4872 is trying to achieve. I thought it would be great to share it here too.

**Comment by taion** [comment]
We actually use them internally in e.g. https://github.com/react-bootstrap/react-bootstrap/blob/v0.30.1/src/DropdownButton.js#L26-L27, so ATM this isn't possible for all the components – not via the Babel transform anyway (and e.g. that's why create-react-app doesn't strip propTypes either).

It'd be nice to get rid of them but it's not clear to me that there's a good path forward that doesn't violate DRY.

**Comment by oliviertassinari** [comment]
> that doesn't violate DRY.
> We actually use them internally in e.g.

I see, you are using the `propTypes` definitions to forward the properties to the right components.

> It's not clear to me that there's a good path forward that doesn't violate DRY.

Wouldn't changing the signature of [`splitComponentProps`](https://github.com/react-bootstrap/react-bootstrap/blob/master/src/utils/splitComponentProps.js#L1) from:
```js
export default function splitComponentProps(props, Component) {
```
to
```js
export default function splitComponentProps(props, ComponentPropTypes) {
```
solve the isssue?
(We would have to export and import the `propTypes`, which would also be more explict)

**Comment by taion** [comment]
It's possible – it'd need to not break how we detect propTypes for the auto-generated docs though.

**Comment by oliviertassinari** [comment]
I guess you are using `react-docgen`. We are using it on Material-UI too. To be confirmed but I'm pretty sure he is supporting it.

**Comment by jquense** [comment]
also The various style utils depend on runtime propTypes as well. that might be fixable but I think this sort of optimization is gonna be lower on the to do list as the bytes saved are minimal

**Comment by taion** [comment]
uncontrollable also has a hard dep on runtime prop type, right?

**Comment by jquense** [comment]
not necessarily, they are used to provided better warnings when missing handlers while controlling props but could work with out then. I think right now if the underlying component has no propTypes defined it'll work fine, albeit with less warnings

**Comment by twhid** [comment]
bumping this issue

Have you considered doing as the [React devs suggest](https://facebook.github.io/react/warnings/dont-call-proptypes.html#dont-call-proptypes-directly) and replacing `React.propTypes` with your own fork? 

> If you depend on using PropTypes like this, we encourage you to use or create a fork of PropTypes [...]

Are there any plans to fix this issue? Am I right in the understanding that anyone depending on this lib is hosed once react@16 drops?

**Comment by taion** [comment]
That's not what that warning guards against, and that's not what this issue is talking about.

**Comment by twhid** [comment]
Thanks for the reply @taion -- you'd like me to open an new issue or PR? Currently, on my server I'm seeing lots of this sort of warning:

> Warning: You are manually calling a React.PropTypes validation function for the `labelledBy` prop on `DropdownMenu`. This is deprecated and will not work in the next major version. You may be seeing this warning due to a third-party PropTypes library. See https://fb.me/react-warning-dont-call-proptypes for details.

Which pretty clearly states that "DropdownMenu" as calling React.propTypes directly. Further, the React devs:
> If you don't fix the warning, this code will crash in production with React 16.

But perhaps I'm not understanding your reply?

**Comment by taion** [comment]
That was fixed almost a _month_ ago in v0.30.1, so no, please do not open a duplicate issue.

**Comment by twhid** [comment]
Great! Our dep was behind. Thanks.
